homeassistant/components/amcrest/camera.py

```diff
@@ -168,18 +265,177 @@ def turn_on(self):
         """Turn on camera."""
         self._enable_video_stream(True)
 
-    # Utility methods
+    def enable_motion_detection(self):
+        """Enable motion detection in the camera."""
+        self._enable_motion_detection(True)
+
+    def disable_motion_detection(self):
+        """Disable motion detection in camera."""
+        self._enable_motion_detection(False)
+
+    # Additional Amcrest Camera service methods
+
+    async def async_enable_recording(self, call_data):
```

Comment:
It's probably better if the entity doesn't need to know about the call data. This method should just take whatever info it needs.

Reply:
The knowledge of the service, from the schema to the method that implements it, needs to be somewhere.

Originally it was all in `camera.py`. Then I moved much of it to `__init__.py`, leaving just the camera methods, with no service knowledge, in `camera.py`. But that required the "hub" code to have too much knowledge (and direct access to) the camera entities.

So then you had me move much of the knowledge back to `camera.py`, leaving just the service registration and processing in `__init__.py`, and I did that in such a way as to decouple the two as much as possible. That means most of the camera specific knowledge is now in `camera.py`, including the service schema and the extraction of the service data. To me, that makes the most sense, since it's all camera specific, and keeps it all in one place.

But now we still have the service registration and initial service handling, including the permission checking, in `__init__.py`, with everything else in `camera.py.` And what is the benefit? I don't see it. In fact, it seems to do nothing but add overhead, specifically the dispatching mechanism. Now when a service is called, not only is there the normal service processing, there's also an extra layer of dispatching, with the dispatcher being clogged up with many more events that don't seem to add any benefit, just extra processing and delay.

At this point I would propose moving the service registration and handling back to `camera.py` so everything is in one place. And I could go back to the way it was originally where the handler extracted the data from the call, which is what you're basically asking for here. The extraction of the entity ids and permission checking can go to the new `helpers.py` module. (In fact, if the code in helpers.service.entity_service_call were refactored, taking an entity id list instead of platforms, I could use it directly and not have to mostly duplicate it.)

Reply:
Ok, done.